NumLib/TimeStepping/Algorithms/EvolutionaryPIDcontroller.cpp

```diff
@@ -0,0 +1,175 @@
+                                _kD) *
+                            h_n;
+                else
+                    h_new = std::pow(_e_n_minus1 / e_n, _kP) *
+                            std::pow(_tol / e_n, _kI) * h_n;
+            }
+            else
+            {
+                h_new = std::pow(_tol / e_n, _kI) * h_n;
+            }
+        }
+
+        h_new = limitStepSize(h_new, h_n);
+        const double checked_h_new = checkSpecificTimeReached(h_new);
+        _dt_vector.push_back(checked_h_new);
+
+        _ts_prev = _ts_current;
+        _ts_current += checked_h_new;
+
+        _e_n_minus2 = _e_n_minus1;
+        _e_n_minus1 = e_n;
+    }
+
+    return true;
+}
+
+double EvolutionaryPIDcontroller::checkSpecificTimeReached(const double h_new)
+{
+    if (_specific_times.empty())
+        return h_new;
+
+    const double specific_time = _specific_times.back();
+    const double zero_threshlod = std::numeric_limits<double>::epsilon();
+    if ((specific_time > _ts_current.current()) &&
+        (_ts_current.current() + h_new - specific_time > zero_threshlod))
+    {
+        _specific_times.pop_back();
+        return specific_time - _ts_current.current();
+    }
+
+    return h_new;
+}
+
+/// Create an EvolutionaryPIDcontroller time stepper from the given
+/// configuration
+std::unique_ptr<TimeStepAlgorithm> createEvolutionaryPIDcontroller(
+    BaseLib::ConfigTree const& config)
+{
+    //! \ogs_file_param{prj__time_loop__time_stepping__type}
+    config.checkConfigParameter("type", "EvolutionaryPIDcontroller");
+
+    //! \ogs_file_param{prj__time_loop__time_stepping__EvolutionaryPIDcontroller__t_initial}
+    auto const t0 = config.getConfigParameter<double>("t_initial");
+    //! \ogs_file_param{prj__time_loop__time_stepping__EvolutionaryPIDcontroller__t_end}
+    auto const t_end = config.getConfigParameter<double>("t_end");
+    //! \ogs_file_param{prj__time_loop__time_stepping__EvolutionaryPIDcontroller__dt_guess}
+    auto const h0 = config.getConfigParameter<double>("dt_guess");
+
+    //! \ogs_file_param{prj__time_loop__time_stepping__EvolutionaryPIDcontroller__dt_min}
+    auto const h_min = config.getConfigParameter<double>("dt_min");
+    //! \ogs_file_param{prj__time_loop__time_stepping__EvolutionaryPIDcontroller__dt_max}
+    auto const h_max = config.getConfigParameter<double>("dt_max");
+    //! \ogs_file_param{prj__time_loop__time_stepping__EvolutionaryPIDcontroller__rel_dt_min}
+    auto const rel_h_min = config.getConfigParameter<double>("rel_dt_min");
+    //! \ogs_file_param{prj__time_loop__time_stepping__EvolutionaryPIDcontroller__rel_dt_max}
+    auto const rel_h_max = config.getConfigParameter<double>("rel_dt_max");
+
+    auto specific_times_opt =
+        //! \ogs_file_param{prj__time_loop__time_stepping__EvolutionaryPIDcontroller__specific_times}
+        config.getConfigParameterOptional<std::vector<double>>(
+            "specific_times");
+    std::vector<double> dummy_vector;
```

Comment:
In the linked [documentation](https://doxygen.opengeosys.org/df/de1/classBaseLib_1_1ConfigTree.html#aed255afa7c8ed1cdcf087593d2251002) it says that
> Get parameter param of type T from the configuration tree or the default_value.
>
> This method has a similar behaviour as getConfigParameter(std::string const&) except in case of errors the default_value is returned.

So in my understanding calling getConfigParameter<vector>("times", vector{}) returns an empty vector, not a null pointer.